Adam and Transformers AdamW CPU solvers upgrade

COrruDXC opened the pull request:
## Motivation

This PR adds two new versions of the Adam and Transformers AdamW CPU solvers to the MIOpen driver.
They are both modified versions of the original solver.

## Technical Details

The updated solver has been optimized for speed, and a multithreaded version of it has also been implemented.
The driver code has been modified to run all the available CPU solvers in a row automatically, without the need of any addidional command line parameter.

## Test Plan

The driver and the GTest related source code have been instrumented to measure (and save to a file) the execution time of each solver while being tested on bench. Such instrumentation is not present in this PR.

## Test Result

Performance depends on the input tensor size, but in most of the cases the modifed versions perform better than the original.
Detailed timings have been attached to this PR.
[Performance tests result.zip](https://github.com/user-attachments/files/24494449/Performance.tests.result.zip)
<img width="3811" height="2467" alt="Adam" src="https://github.com/user-attachments/assets/ed44b394-8633-44df-8229-9ef67eba770b" />
<img width="3564" height="2247" alt="TransformersAdam" src="https://github.com/user-attachments/assets/1528eaed-8684-44f4-a7e7-1df10a5b0232" />





## Submission Checklist

- [X] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

SreecharanGundaboluAMD commented:
Hey, could you please revert the Clang-format fix for unrelated filed that are not part of your change. Also, please rebase your branch as the latest develop already addresses these clang format issues.

ScottTodd commented on .github/scripts/therock_matrix.py:
Can you split this into a few smaller PRs? This PR is touching quite a few files across three subprojects and some continuous integration systems. From CODEOWNERS, there are 13 direct review requests and 3 team review requests. If you split this into several smaller PRs they will be easier to review.

In particular, I think the code changes to e.g. `projects/miopen/driver/adam_driver.hpp` look independent from the changes in `.github/scripts/therock_matrix.py`, or at least they could be sequenced one after another during review. If the CI (`.github/*`) changes are useful/necessary for testing the changes prior to first landing them, they might be okay to include here.

BrianHarrisonAMD replied:
Looks like a strange rebase issue.
I think a lot of these changes landed last week from separate PRs unrelated to this work.
(Just noticed since I can see these duplicating changes found in #2919)

SreecharanGundaboluAMD replied:
@COrruDXC It looks like a rebase issue is causing duplicate commits to appear in your PR, which are unrelated to the files in your changes. We want to ensure that PRs only include files relevant to your work and avoid any unrelated changes within miopen or other projects. Could you please close this PR and create a new one from a fresh branch, without the clang-format commit or the faulty rebase? This will help us merge cleanly and keep the commit history tidy when we squash and merge.

COrruDXC replied:
Yes, it's definitely a rebase issue. I'm working to fix it.

COrruDXC replied:
The rebase issue should be fixed now. Since I wanted to avoid creating a new PR, I synchronized my branch with the upstream repository. And now the 'Files changed' tab only shows my modifications.
The PR is now ready for review.

SreecharanGundaboluAMD commented on projects/miopen/driver/adam_driver.hpp:
```diff
@@ -107,17 +115,94 @@ void mloAdamRunHost(miopenTensorDescriptor_t paramDesc,
+    const float one_minus_beta1              = 1.0 - beta1;
+    const float one_minus_beta2              = 1.0 - beta2;
+
+    miopen::par_ford(numel)([&](int32_t i) {
```
There appears to be significant code duplication between the single-threaded and multi-threaded implementations. To improve maintainability, we want to reduce this duplication. [Here is a discussion](https://github.com/ROCm/rocm-libraries/pull/2194#discussion_r2557724285) on the topic, please review it and make the necessary changes to address the issue. We understand that even with the best approach, there may be some performance impact. Could you also collect and share the raw performance numbers to help us assess the impact of these changes?

SreecharanGundaboluAMD commented on projects/miopen/driver/adam_driver.hpp:
```diff
@@ -571,15 +684,35 @@ int AdamDriver<Tgpu, Tref, Tgrad>::VerifyForward()
+
+        if(!std::isfinite(error) || error > tolerance)
+        {
+            std::cout << "Forward Adam FAILED on updated multi-threaded CPU reference: " << error
```
What is the need for the error to say "updated" test failed? I think we can remove that across all the error prints.

SreecharanGundaboluAMD commented on projects/miopen/driver/adam_driver.hpp:
```diff
@@ -571,15 +684,35 @@ int AdamDriver<Tgpu, Tref, Tgrad>::VerifyForward()
+
+        std::cout << "Forward Adam Verifies OK on updated single-threaded CPU reference"
+                  << std::endl;
+    }
```
There seems to be duplication here as well, consider something like this for logging.

```suggestion
    const auto error = miopen::rms_range(param_host, param);
    const char* ref_type =
        use_multithread ? "multi-threaded" : "single-threaded";

    if(!std::isfinite(error) || error > tolerance)
    {
        std::cout << "Forward Adam FAILED on " << ref_type
                  << " CPU reference: " << error << std::endl;
        return EC_VerifyFwd;
    }

    std::cout << "Forward Adam Verifies OK on " << ref_type
              << " CPU reference" << std::endl;
```

SreecharanGundaboluAMD commented on projects/miopen/driver/transformers_adam_w_driver.hpp:
```diff
@@ -398,45 +404,99 @@ int TransformersAdamWDriver<Tgpu, Tref, Tgrad>::RunForwardCPU()
+}
+
+template <typename Tgpu, typename Tref, typename Tgrad>
+int TransformersAdamWDriver<Tgpu, Tref, Tgrad>::RunForwardCPU_MT()
```
Same comment here, kindly address the duplication issue.

SreecharanGundaboluAMD commented on projects/miopen/driver/transformers_adam_w_driver.hpp:
```diff
@@ -78,12 +78,15 @@ class TransformersAdamWDriver : public Driver
 
     int RunForwardGPU() override;
     int RunForwardCPU();
+    int RunForwardCPUUpdated();
```
This method is not defined or used elsewhere in your PR.

SreecharanGundaboluAMD commented on projects/miopen/driver/transformers_adam_w_driver.hpp:
```diff
@@ -472,17 +532,40 @@ Tref TransformersAdamWDriver<Tgpu, Tref, Tgrad>::GetTolerance()
+
+        std::cout << "Forward Transformers Adam Verifies OK on single-threaded CPU reference"
+                  << std::endl;
+    }
```
Reduce duplication.

SreecharanGundaboluAMD commented on projects/miopen/test/gtest/adam.cpp:
```diff
@@ -60,6 +98,109 @@ TEST_P(GPU_AmpAdam_FP32, AmpAdamTestFw)
-INSTANTIATE_TEST_SUITE_P(Full, GPU_Adam_FP32, testing::ValuesIn(AdamTestConfigs()));
-INSTANTIATE_TEST_SUITE_P(Full, GPU_Adam_FP16, testing::ValuesIn(AdamTestConfigs()));
-INSTANTIATE_TEST_SUITE_P(Full, GPU_AmpAdam_FP32, testing::ValuesIn(AdamTestConfigs()));
+/////////////////////////////////////////////////////////
```
Could you please elaborate on what you’re trying to accomplish here? From what I understand, you’re testing new implementations of the parallel and single-threaded Adam algorithms against the existing ones, is that correct? If the goal is purely to validate your implementation, and assuming it’s thoroughly tested and performs better, it would eventually replace the original implementation. In that case, there’s no need to keep both the original and updated versions in the codebase at the same time.

SreecharanGundaboluAMD commented on projects/miopen/test/gtest/adam.hpp:
```diff
@@ -253,6 +253,154 @@ struct AdamTest : public ::testing::TestWithParam<AdamTestCase>
+        const miopen::TensorDescriptor emptyDesc;
+        auto&& handle = get_handle();
+
+        cpu_adam_updated<Tp, Tg>(ref_param,
```
The names for these methods are confusing. It seems that `cpu_adam_updated` is a parallel implementation, while you explicitly name the single-threaded one. I would suggest naming them `cpu_adam_updated` for the single-threaded version and `cpu_adam_updated_multi_threaded` for the multi-threaded version. But if my above comment is accurate, then there is no need for these methods to exist in the first place.

COrruDXC replied:
> The names for these methods are confusing. It seems that `cpu_adam_updated` is a parallel implementation, while you explicitly name the single-threaded one. I would suggest naming them `cpu_adam_updated` for the single-threaded version and `cpu_adam_updated_multi_threaded` for the multi-threaded version. But if my above comment is accurate, then there is no need for these methods to exist in the first place.

I left the names the same way as the original ones.
The original GTest solver was named 'cpu_adam' and it was multi-threaded. That's why I named my updated multi-threaded version 'cpu_adam_updated'. As a consequence, I named 'cpu_adam_updated_singlethread' the single-threaded version based on the updated multi-threaded one.
Anyway, to dispel all doubts, after the code duplication removal only the overall best-performing version will remain.

SreecharanGundaboluAMD commented on projects/miopen/test/gtest/transformers_adam_w.hpp:
```diff
@@ -235,6 +235,168 @@ struct TransformersAdamWTest : public ::testing::TestWithParam<TransformersAdamW
             param_fp16.data = handle.Read<half_float::half>(param_fp16_dev, param_fp16.data.size());
     }
 
+    void RunUpdatedTest()
```
Same comments as the previous.

SreecharanGundaboluAMD reviewed (changes requested):
In the graph that you attached, can you explain why your updated multithreaded implementation is slower than the single threaded reference implementation at larger sizes? Why is the "original" multithreaded implementation not a part of your performance study? Kindly reduce code duplication throughout your changes. Also, if the aim of your PR is to update the existing CPU reference with your updated ones, could you please thoroughly test the implementations against each other and remove the gtests that are comparing the two reference CPU implementations?

COrruDXC commented:
> In the graph that you attached, can you explain why your updated multithreaded implementation is slower than the single threaded reference implementation at larger sizes?

I have no exact answer for this. But I think it's related to the par_ford overhead. From what I could observe, par_ford performs better with long algorithms, because in such case its overhead is negligible compared to the runtime of the parallelized algorithm. But in case of a short or medium length algorithm, the overhead of par_ford becomes comparable to the algorithm runtime, causing a serious performance hit. In this case the single-threaded version performs better because it lacks such overhead.
Also, I ran the tests on our internal workstation equipped with an AMD Ryzen 9 5900X 12-Core/24-Thread Processor. Probably we need more hardware threads to see the multi-threaded version shine.

> Why is the "original" multithreaded implementation not a part of your performance study?

Every original implementation is part of my performance study. But the CPU solver in the driver has no "original multithreaded" implementation. Only GTest has one.
Three versions of the same solver both for the driver and GTest have been tested:

- Driver
        1. Original version: single-threaded
        2. Modified single-threaded version: improved single-threaded version of the original
        3. Modified multi-threaded version: multi-threaded version of the improved single-threaded version

- GTest
        1. Original version: multi-threaded
        2. Modified multi-threaded version: improved version of the original
        3. Modified single-threaded version: single-threaded version of the improved multi-threaded version

The original single-threaded version has been removed from the driver because it was the worst performing one.
The attached graphs refer to the driver versions of the Adam and Transformers AdamW solvers. For the complete results, please see the zip file attached to the PR.

> Kindly reduce code duplication throughout your changes.

I'm working on it.

> Also, if the aim of your PR is to update the existing CPU reference with your updated ones, could you please thoroughly test the implementations against each other and remove the gtests that are comparing the two reference CPU implementations?

GTests are not comparing CPU implementation against each other. Each GTest compares one CPU implementation against the GPU implementation.
After removing the code duplication, there will be only the overall best-performing solver version left in the GTest part.

marbre commented on projects/miopen/driver/adam_driver.hpp:
```diff
@@ -2,7 +2,7 @@
  * MIT License
  *
- * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ * Copyright (c) 2025 Advanced Micro Devices, Inc.
```
Drive-by comment: Don't update the year here. This is the year when the file was released and if it isn't a date range, do not update.

SreecharanGundaboluAMD commented on projects/miopen/driver/transformers_adam_w_driver.hpp:
```diff
@@ -472,17 +479,27 @@ Tref TransformersAdamWDriver<Tgpu, Tref, Tgrad>::GetTolerance()
+    RunForwardCPU();
+    const auto t1 = std::chrono::high_resolution_clock::now();
+    const auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();
+    std::ofstream ofs("/data/Dev/MIOpenDriverStats-" + this->name + ".csv", std::ofstream::app);
```
I guess this is supposed to be here only for debugging purposes, could you please remove this.

SreecharanGundaboluAMD commented on projects/miopen/driver/adam_driver.hpp:
```diff
@@ -48,54 +49,63 @@
+                       bool found_inf,
+                       bool multi_threaded)
 {
+    const auto t0 = std::chrono::high_resolution_clock::now();
```
I am assuming that this debugging code is going to be cleaned.

SreecharanGundaboluAMD commented on projects/miopen/driver/adam_driver.hpp:
```diff
@@ -48,54 +49,63 @@
+    const float one_minus_beta2              = 1.0f - beta2;
+    const size_t min_grain                   = multi_threaded ? 8 : numel;
+
+    miopen::par_for(numel, min_grain, [&](int i) {
```
Nice efforts to reduce code duplication. Just curious is there any performance impact that was observed between the simple for loop and par_ford algorithm for single threaded behavior?

COrruDXC replied:
In general the single-threaded par_for based solver is slower than its plain for loop based counterpart.

SreecharanGundaboluAMD reviewed (approved):
LGTM
